content/blog/2022-06-01-Jupyter-in-cloud-with-TPI.md

```diff
@@ -0,0 +1,379 @@
+    (curl -fsSL https://deb.nodesource.com/setup_16.x | bash -) >/dev/null
+    apt-get install -y $${QUIET:+-qq} nodejs
+
+    # start tunnel
+    export JUPYTER_TOKEN="$(uuidgen)"
+    pushd "$(mktemp -d --suffix dependencies)"
+    npm i ngrok
+    npx ngrok authtoken "$NGROK_TOKEN"
+    (node <<TUNNEL
+    const fs = require('fs');
+    const ngrok = require('ngrok');
+    (async function() {
+      const jupyter = await ngrok.connect(8888);
+      const tensorboard = await ngrok.connect(6006);
+      const br = '\n*=*=*=*=*=*=*=*=*=*=*=*=*\n';
+      fs.writeFileSync("log.md", \`\$${br}URL: Jupyter Lab: \$${jupyter}/lab?token=$${JUPYTER_TOKEN}\$${br}URL: Jupyter Notebook: \$${jupyter}/tree?token=$${JUPYTER_TOKEN}\$${br}URL: TensorBoard: \$${tensorboard}\$${br}\`);
+    })();
+    TUNNEL
+    ) &
+    while test ! -f log.md; do sleep 1; done
+    cat log.md
+    popd # dependencies
+
+    # start tensorboard in background
+    env -u JUPYTER_TOKEN -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN tensorboard --logdir . --host 0.0.0.0 --port 6006 &
+
+    # start Jupyter server in foreground
+    env -u AWS_ACCESS_KEY_ID -u AWS_SECRET_ACCESS_KEY -u REPO_TOKEN jupyter lab --allow-root --ip=0.0.0.0 --no-browser --port=8888 --port-retries=0
+  END
+}
+```
+
+The second-to-last part of `main.tf` contains the script that will run once the
+instance has been provisioned. These are CLI commands as you would run them in
+your local terminal. We won't discuss every individual line, but in broad
+strokes, the script does the following:
+
+1. Run a few configuration commands
+1. Install the dependencies we will be needing
+1. Start an ngrok tunnel so that we can easily access the server through our
+   browser
+1. Launch the Jupyter server
+
+<admon type="info">
+
+This script also launches
+[Tensorboard](https://www.tensorflow.org/tensorboard/get_started) to run
+alongside Jupyter. It's not the main focus of this blog post, but it might be
+helpful!
+
+If you don't want Tensorboard running, simply remove or comment out
+[line 78 of `main.tf`](https://github.com/iterative/blog-tpi-jupyter/blob/aws/main.tf#L78).
+
+</admon>
+
+As you can see, the possibilities for the scripts we can run through TPI are
+extensive. We could add commands to clone a Git repository, for example. Or we
+could pull data in from a [DVC](https://dvc.org/) remote. This flexibility
+allows us to tailor the instance TPI provisions precisely to our needs.
+
+```bash
+output "urls" {
+  value = flatten(regexall("URL: (.+)", try(join("\n", iterative_task.jupyter_server.logs), "")))
+}
+```
+
+Lastly, we specify the outputs for our script. This is the value Terraform
+returns to be shown in our terminal after provisioning our instance. In this
+case, we are returning the URLs through which we can access the Jupyter server.
+
+# Next steps: rethink this approach
```

Comment:
this misses the point of the blog I think. It's not about "is Jupyter a good idea" it's more about "bring-your-own-cloud is cheaper and more configurable than depending on a comprehensive package solution."

See overall comments.